File: Sources/ConnectorControlCore/MasterStore.swift
import Foundation

public enum EditView: String, Codable, Hashable, Sendable {
    case form, json
}

public struct MCPEntry: Equatable, Hashable, Codable, Sendable {
    public var enabled: Bool
    public var config: JSONValue
    public var lastEditView: EditView

    public init(enabled: Bool = true, config: JSONValue, lastEditView: EditView = .form) {
        self.enabled = enabled
        self.config = config
        self.lastEditView = lastEditView
    }
}

/// A full, independent snapshot of connectors: its own configs + enabled flags.
public struct Collection: Equatable, Codable, Sendable {
    public var mcps: [String: MCPEntry]
    public init(mcps: [String: MCPEntry] = [:]) { self.mcps = mcps }
}

/// Schema v2 only — no v1 fallback. A v1 (or otherwise malformed) file on
/// disk fails to decode and is handled by `MasterStoreIO.load`'s existing
/// corrupt-file path: moved aside and rebuilt fresh from Claude's config.
public struct MasterStore: Equatable, Codable, Sendable {
    public var version: Int
    public var activeCollection: String
    public var collections: [String: Collection]

    // The file keeps the v2 key names: machines on the current release share it through the
    // synced master-list folder, and their decoder knows only these two keys.
    enum CodingKeys: String, CodingKey {
        case version
        case activeCollection = "activeProfile"
        case collections = "profiles"
    }

    /// The active collection's connectors — the view the entire app operates on.
    public var mcps: [String: MCPEntry] {
        get { collections[activeCollection]?.mcps ?? [:] }
        set { collections[activeCollection, default: Collection()].mcps = newValue }
    }

    /// Claude's `mcpServers` section rendered from this store — the enabled
    /// subset's configs. The store is the source of truth; Claude's config is
    /// downstream, and any divergence from this render is regenerated away.
    public var enabledServers: [String: JSONValue] {
        mcps.filter(\.value.enabled).mapValues(\.config)
    }

    public static let empty = MasterStore(
        activeCollection: "Default",
        collections: ["Default": Collection()])

    public init(activeCollection: String, collections: [String: Collection]) {
        self.version = 2
        self.activeCollection = activeCollection
        self.collections = collections
    }

    /// The name a collection is kept under for the one typed: spaces trimmed from both ends.
    /// Adding and renaming both apply it, so a caller that follows the collection it just named
    /// asks here rather than trimming again.
    public static func collectionName(_ typed: String) -> String { typed.trimmingCharacters(in: .whitespaces) }

    /// nil on success, else a user-facing error message. The new collection becomes the active
    /// one unless `activating` is false.
    public mutating func addCollection(named name: String, copyingCurrent: Bool, activating: Bool = true) -> String? {
        let trimmed = Self.collectionName(name)
        guard !trimmed.isEmpty else { return "Name must not be empty." }
        guard collections[trimmed] == nil else {
            return "A collection named \u{201C}\(trimmed)\u{201D} already exists."
        }
        collections[trimmed] = copyingCurrent ? Collection(mcps: mcps) : Collection()
        if activating { activeCollection = trimmed }
        return nil
    }

    /// nil on success, else a user-facing error message. Renaming the active
    /// collection keeps it active under its new name.
    public mutating func renameCollection(_ name: String, to newName: String) -> String? {
        guard collections[name] != nil else { return Self.noCollectionError(name) }
        let trimmed = Self.collectionName(newName)
        guard !trimmed.isEmpty else { return "Name must not be empty." }
        if trimmed != name, collections[trimmed] != nil {
            return "A collection named \u{201C}\(trimmed)\u{201D} already exists."
        }
        guard let current = collections.removeValue(forKey: name) else { return nil }
        collections[trimmed] = current
        if activeCollection == name { activeCollection = trimmed }
        return nil
    }

    /// nil on success, else a user-facing error message. Refuses to delete the
    /// last remaining collection. Deleting the active collection hands the
    /// sorted-first remaining collection the active spot.
    public mutating func deleteCollection(named name: String) -> String? {
        guard collections[name] != nil else { return Self.noCollectionError(name) }
        guard collections.count > 1 else { return "Can\u{2019}t delete the last collection." }
        let successor = activeAfterDeleting(name)
        collections.removeValue(forKey: name)
        if activeCollection == name { activeCollection = successor ?? "Default" }
        return nil
    }

    /// The collection that takes the active spot if `name` is deleted: the sorted-first of the
    /// rest, or nil when none remain. The one rule, so the Delete confirmation that names it
    /// cannot disagree with the delete that picks it.
    public func activeAfterDeleting(_ name: String) -> String? {
        collections.keys.filter { $0 != name }.min { $0.ordinallyPrecedes($1) }
    }

    /// The one wording for a name no collection has, shared by switch, rename and delete.
    static func noCollectionError(_ name: String) -> String { "No collection named \u{201C}\(name)\u{201D}." }

    public mutating func switchCollection(to name: String) -> String? {
        guard collections[name] != nil else { return Self.noCollectionError(name) }
        activeCollection = name
        return nil
    }
}

public enum MasterStoreIO {
    /// Missing file → empty store. Corrupt file → moved aside to
    /// `mcps.corrupt.<timestamp>.json` (returned) and an empty store; the caller
    /// repopulates it by reconciling against Claude's config.
    public static func load(
        from url: URL, now: Date = Date()
    ) -> (store: MasterStore, corruptFileURL: URL?) {
        let fm = FileManager.default
        guard fm.fileExists(atPath: url.path) else { return (.empty, nil) }
        do {
            let data = try Data(contentsOf: url)
            var store = try JSONDecoder().decode(MasterStore.self, from: data)
            // Self-heal a decoded-but-inconsistent activeCollection (hand-edited
            // or corrupted file) — never crash; fall back to an existing
            // collection (sorted first), or a fresh Default if none remain.
            if store.collections[store.activeCollection] == nil {
                if let fallback = store.collections.keys.sorted().first {
                    store.activeCollection = fallback
                } else {
                    store.collections["Default"] = Collection()
                    store.activeCollection = "Default"
                }
            }
            return (store, nil)
        } catch {
            let stamp = BackupTimestamp.string(from: now)
            let aside = url.deletingLastPathComponent()
                .appendingPathComponent("mcps.corrupt.\(stamp).json")
            do {
                try fm.moveItem(at: url, to: aside)
                return (.empty, aside)
            } catch {
                // Couldn't move it aside; the corrupt file stays in place.
                return (.empty, url)
            }
        }
    }

    public static func save(_ store: MasterStore, to url: URL, staging: URL? = nil) throws {
        try AtomicFile.write(try JSONEncoder.canonical.encode(store), to: url, staging: staging)
    }

    /// Side-effect-free peek: nil when the file is missing or undecodable.
    /// Unlike `load`, never moves a corrupt file aside — used by the store
    /// watcher to classify an on-disk change (own write echo, external edit,
    /// or a sync tool's mid-write partial) before deciding to adopt it.
    public static func read(from url: URL) -> MasterStore? {
        guard let data = try? Data(contentsOf: url) else { return nil }
        return try? JSONDecoder().decode(MasterStore.self, from: data)
    }
}

File: windows/src/ConnectorControl.Core/MasterStore.cs
namespace ConnectorControl.Core;

/// <summary>
/// The master list (schema v2, collection-aware). Mutable like the Swift struct's
/// `var` usage; structural equality like the Swift value type. Use
/// <see cref="Clone"/> where Swift relied on copy semantics.
/// </summary>
public sealed class MasterStore : IEquatable<MasterStore>
{
    // Schema v2 only — no v1 fallback. A v1 (or otherwise malformed) file on
    // disk fails to decode and is handled by MasterStoreIO.Load's existing
    // corrupt-file path: moved aside and rebuilt fresh from Claude's config.
    public const long CurrentVersion = 2;

    public long Version { get; }
    public string ActiveCollection { get; set; }
    public Dictionary<string, Collection> Collections { get; }

    public MasterStore(long version, string activeCollection, IEnumerable<KeyValuePair<string, Collection>> collections)
    {
        Version = version;
        Collections = new Dictionary<string, Collection>(collections, StringComparer.Ordinal);
        // The Mcps getter must never create a collection as a side effect, so this constructor is
        // the one place that guarantees Collections[ActiveCollection] exists — including a decoded
        // file that names a collection it doesn't have (MasterStoreIO.Load relies on this).
        if (Collections.ContainsKey(activeCollection))
        {
            ActiveCollection = activeCollection;
        }
        else if (Collections.Keys.Order(StringComparer.Ordinal).FirstOrDefault() is { } fallback)
        {
            ActiveCollection = fallback;
        }
        else
        {
            Collections["Default"] = new Collection();
            ActiveCollection = "Default";
        }
    }

    /// <summary>Swift <c>MasterStore(version:mcps:)</c>: a single "Default" collection; always v2.</summary>
    public MasterStore(IEnumerable<KeyValuePair<string, McpEntry>> mcps)
        : this(CurrentVersion, "Default", [new KeyValuePair<string, Collection>("Default", new Collection(mcps))])
    {
    }

    /// <summary>Swift <c>.empty</c>. A fresh instance every call — this type is mutable.</summary>
    public static MasterStore Empty() =>
        new(CurrentVersion, "Default", [new KeyValuePair<string, Collection>("Default", new Collection())]);

    /// <summary>
    /// The active collection's connectors — the view the entire app operates on. Side-effect free:
    /// an active collection that somehow doesn't exist yet returns an empty, unstored dictionary
    /// rather than creating one — the constructor is what normally guarantees
    /// Collections[ActiveCollection] exists.
    /// </summary>
    public Dictionary<string, McpEntry> Mcps =>
        Collections.TryGetValue(ActiveCollection, out var collection) ? collection.Mcps : new Dictionary<string, McpEntry>(StringComparer.Ordinal);

    /// <summary>Claude's <c>mcpServers</c> section rendered from this store: the enabled subset's configs.</summary>
    public IReadOnlyDictionary<string, JsonValue> EnabledServers =>
        Mcps.Where(p => p.Value.Enabled).ToDictionary(p => p.Key, p => p.Value.Config, StringComparer.Ordinal);

    /// <summary>How many connectors are enabled, without building the config dictionary <see cref="EnabledServers"/> does.</summary>
    public int EnabledCount => Mcps.Count(p => p.Value.Enabled);

    /// <summary>
    /// The name a collection is kept under for the one typed: spaces trimmed from both ends.
    /// Adding and renaming both apply it, so a caller that follows the collection it just named
    /// asks here rather than trimming again.
    /// </summary>
    public static string CollectionName(string typed) => typed.TrimSpaces();

    /// <summary>
    /// null on success, else a user-facing error message. The new collection becomes the active
    /// one unless <paramref name="activating"/> is false.
    /// </summary>
    public string? AddCollection(string name, bool copyingCurrent, bool activating = true)
    {
        var trimmed = CollectionName(name);
        if (trimmed.Length == 0)
        {
            return "Name must not be empty.";
        }
        if (Collections.ContainsKey(trimmed))
        {
            return $"A collection named “{trimmed}” already exists.";
        }
        Collections[trimmed] = copyingCurrent ? new Collection(Mcps) : new Collection();
        if (activating)
        {
            ActiveCollection = trimmed;
        }
        return null;
    }

    /// <summary>null on success, else a user-facing error message. Renaming the active collection keeps it active under its new name.</summary>
    public string? RenameCollection(string name, string newName)
    {
        if (!Collections.ContainsKey(name))
        {
            return NoCollectionError(name);
        }
        var trimmed = CollectionName(newName);
        if (trimmed.Length == 0)
        {
            return "Name must not be empty.";
        }
        if (trimmed != name && Collections.ContainsKey(trimmed))
        {
            return $"A collection named “{trimmed}” already exists.";
        }
        if (!Collections.Remove(name, out var current))
        {
            return null;
        }
        Collections[trimmed] = current;
        if (ActiveCollection == name)
        {
            ActiveCollection = trimmed;
        }
        return null;
    }

    /// <summary>
    /// null on success, else a user-facing error message. Refuses to delete the last remaining
    /// collection. Deleting the active collection hands the sorted-first remaining collection
    /// the active spot.
    /// </summary>
    public string? DeleteCollection(string name)
    {
        if (!Collections.ContainsKey(name))
        {
            return NoCollectionError(name);
        }
        if (Collections.Count <= 1)
        {
            return "Can’t delete the last collection.";
        }
        var successor = ActiveAfterDeleting(name);
        Collections.Remove(name);
        if (ActiveCollection == name)
        {
            ActiveCollection = successor ?? "Default";
        }
        return null;
    }

    /// <summary>
    /// The collection that takes the active spot if <paramref name="name"/> is deleted: the
    /// sorted-first of the rest, or null when none remain. The one rule, so the Delete
    /// confirmation that names it cannot disagree with the delete that picks it.
    /// </summary>
    public string? ActiveAfterDeleting(string name) =>
        Collections.Keys.Where(k => k != name).Order(StringComparer.Ordinal).FirstOrDefault();

    /// <summary>The one wording for a name no collection has, shared by switch, rename and delete.</summary>
    private static string NoCollectionError(string name) => $"No collection named “{name}”.";

    public string? SwitchCollection(string name)
    {
        if (!Collections.ContainsKey(name))
        {
            return NoCollectionError(name);
        }
        ActiveCollection = name;
        return null;
    }

    public MasterStore Clone() =>
        new(Version, ActiveCollection, Collections.Select(p => new KeyValuePair<string, Collection>(p.Key, p.Value.Clone())));

    // MARK: JSON (the Swift Codable synthesis, made explicit)

    // The file keeps the v2 key names: machines on the current release share it through the
    // synced master-list folder, and their decoder knows only these two keys.
    public JsonValue ToJson() => JsonValue.Object(
        ("version", JsonValue.Int(Version)),
        ("activeProfile", JsonValue.String(ActiveCollection)),
        ("profiles", JsonValue.Object(Collections.Select(p =>
            new KeyValuePair<string, JsonValue>(p.Key, CollectionToJson(p.Value))))));

    private static JsonValue CollectionToJson(Collection collection) => JsonValue.Object(
        ("mcps", JsonValue.Object(collection.Mcps.Select(m =>
            new KeyValuePair<string, JsonValue>(m.Key, EntryToJson(m.Value))))));

    private static JsonValue EntryToJson(McpEntry entry) => JsonValue.Object(
        ("enabled", JsonValue.Bool(entry.Enabled)),
        ("config", entry.Config),
        ("lastEditView", JsonValue.String(entry.LastEditView.ToJsonString())));

    /// <summary>
    /// Strict like Swift's synthesized Codable: every key required with the
    /// right type, unknown keys ignored. Throws <see cref="FormatException"/>.
    /// </summary>
    public static MasterStore FromJson(JsonValue json)
    {
        if (json.Kind != JsonKind.Object)
        {
            throw new FormatException("master store: top level is not an object");
        }
        // The file keeps the v2 key names: machines on the current release share it through the
        // synced master-list folder, and their decoder knows only these two keys.
        var active = Require(json, "activeProfile", JsonKind.String).StringValue;
        var collections = Require(json, "profiles", JsonKind.Object).ObjectProperties
            .Select(p => new KeyValuePair<string, Collection>(p.Key, CollectionFromJson(p.Value)));
        return new MasterStore(Require(json, "version", JsonKind.Int).IntValue, active, collections);
    }

    private static Collection CollectionFromJson(JsonValue json)
    {
        if (json.Kind != JsonKind.Object)
        {
            throw new FormatException("master store: collection is not an object");
        }
        var mcps = Require(json, "mcps", JsonKind.Object).ObjectProperties
            .Select(m => new KeyValuePair<string, McpEntry>(m.Key, EntryFromJson(m.Value)));
        return new Collection(mcps);
    }

    private static McpEntry EntryFromJson(JsonValue json)
    {
        if (json.Kind != JsonKind.Object)
        {
            throw new FormatException("master store: entry is not an object");
        }
        var enabled = Require(json, "enabled", JsonKind.Bool).BoolValue;
        var config = json["config"] ?? throw new FormatException("master store: entry.config missing");
        var rawView = Require(json, "lastEditView", JsonKind.String).StringValue;
        if (!EditViewJson.TryParse(rawView, out var view))
        {
            throw new FormatException($"master store: unknown lastEditView '{rawView}'");
        }
        return new McpEntry(enabled, config, view);
    }

    private static JsonValue Require(JsonValue obj, string key, JsonKind kind)
    {
        var value = obj[key] ?? throw new FormatException($"master store: '{key}' missing");
        if (value.Kind != kind)
        {
            throw new FormatException($"master store: '{key}' is {value.TypeName}, expected {kind}");
        }
        return value;
    }

    // MARK: equality

    public bool Equals(MasterStore? other) =>
        other is not null
        && Version == other.Version
        && string.Equals(ActiveCollection, other.ActiveCollection, StringComparison.Ordinal)
        && DictionaryEquality.Equal(Collections, other.Collections);

    public override bool Equals(object? obj) => Equals(obj as MasterStore);

    public override int GetHashCode() =>
        HashCode.Combine(Version, ActiveCollection.GetHashCode(StringComparison.Ordinal), DictionaryEquality.Hash(Collections));
}
